aten/src/ATen/native/cpu/ActivationKernel.cpp

```diff
@@ -0,0 +1,484 @@
+  }
+}
+
+template <typename scalar_t>
+inline void _vec_sub_exp_scalarmul_write(
+    scalar_t* grad_input_data,
+    scalar_t* grad_data,
+    scalar_t* output_data,
+    int64_t dim_size,
+    scalar_t sum) {
+  using Vec = vec256::Vec256<scalar_t>;
+  Vec sum_vec(sum);
+  int64_t d = 0;
+  for (; d < dim_size; d += Vec::size) {
+    Vec output(0);
+    Vec grad(0);
+    int64_t leftover = _leftover<Vec::size>(d, dim_size);
+    output.load(output_data + d, leftover);
+    grad.load(grad_data + d, leftover);
+    output = output.exp();
+    output = output * sum_vec;
+    grad = grad - output;
+
+    grad.store(grad_input_data + d, leftover);
+  }
+}
+
+template <typename scalar_t>
+inline scalar_t _vec_norm_exp_sum_write(
+    scalar_t* output_data,
+    int64_t dim_size,
+    scalar_t* input_data,
+    scalar_t max_input) {
+  using Vec = vec256::Vec256<scalar_t>;
+
+  int64_t d = 0;
+  Vec tmpsum(0);
+  Vec max_input_vec(max_input);
+  scalar_t tmp_sum_arr[Vec::size];
+  scalar_t tmp_sum_scalar = 0;
+  for (; d < dim_size; d += Vec::size) {
+    int64_t leftover = _leftover<Vec::size>(d, dim_size);
+    Vec value(0);
+    value.load(input_data + d, leftover);
+    value = value - max_input_vec;
+    value = value.exp();
+    // Need to do a partial add
+    if (d + Vec::size > dim_size) {
+      scalar_t value_arr[Vec::size];
+      value.store(value_arr);
+      for (int64_t i = 0; i < dim_size - d; i++) {
+        tmp_sum_scalar += value_arr[i];
+      }
+      value.store(output_data + d, leftover);
+    } else {
+      tmpsum = tmpsum + value;
+      value.store(output_data + d);
+    }
+  }
+  tmpsum.store(tmp_sum_arr);
+  for (int64_t i = 0; i < Vec::size; i++) {
+    tmp_sum_scalar += tmp_sum_arr[i];
+  }
+  return tmp_sum_scalar;
+}
+
+template <typename scalar_t>
+inline void _vec_log_softmax_lastdim(
+    scalar_t* input_data_base,
+    scalar_t* output_data_base,
+    int64_t outer_size,
+    int64_t dim_size) {
+  using Vec = vec256::Vec256<scalar_t>;
+  static constexpr int64_t CHUNK_SIZE = (128 / sizeof(scalar_t)) * Vec::size;
+  int64_t grain_size = internal::TBB_GRAIN_SIZE / (dim_size * CHUNK_SIZE);
+  if (grain_size < CHUNK_SIZE)
+    grain_size = CHUNK_SIZE;
+  //std::cout << "CHUNK_SIZE: " << CHUNK_SIZE<< std::endl;
+  //std::cout << "Vec::size: " << Vec::size << std::endl;
+  //std::cout << "outer_size: " << outer_size<< std::endl;
+
+  tbb::parallel_for(
+      tbb::blocked_range<int64_t>(0, outer_size, grain_size),
+      [&](const tbb::blocked_range<int64_t>& r) {
+        for (int64_t ii = r.begin(); ii < r.end(); ii += CHUNK_SIZE) {
+          scalar_t tmp_sum_scalar[CHUNK_SIZE];
+          scalar_t max_input_arr[CHUNK_SIZE];
+          int64_t inner_loop_end = CHUNK_SIZE;
+          if (ii + CHUNK_SIZE > r.end())
+            inner_loop_end = r.end() - ii;
+          //std::cout<< "inner_loop_end: " << inner_loop_end << std::endl;
+          for (int64_t j = 0; j < inner_loop_end; j++) {
```

Comment:
For the future: IIRC it is slightly better to use unsigned types for loop iterators if you can, because their overflow is well defined, and it compiles to fewer instructions in general.

Reply:
Sure, we can use size_t all the time. I've shot myself in the foot using those before because of overflows (negative values), but maybe it's worth it.

Reply:
Uhhhhh... the opposite is true. Stick with signed integer types. Most of the time it doesn't matter, but it:

a) is easier to be consistent and we use signed integer types most places
b) it enables some strength reduction (see https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#loop-counters-signed-vs-unsigned for an explanation which applies to CPU as well)

Reply:
Uh, that actually makes sense. Thanks @colesbury!